### PR TITLE
first draft

### PR DESCRIPTION
**Pull request by alex000kim**
The first draft of the post titled "End-to-End Computer Vision API with DVC, CML and Studio".
TBD:
- Redo diagrams
- Banner/header picture
- `commentsUrl` for https://discuss.dvc.org

### CODE REVIEW

**Comment by dberenbaum** [comment]
Wow @alex000kim, this is great stuff! Very comprehensive and well done. The content is really well polished for a first draft!

There's a lot going on in this blog post. What would you say is the main objective or focus of the post? It covers so many things, it's not totally clear to me. If that's the point and it is about showing an end-to-end ML project, has there been any consideration of breaking it down into a multi-part series of posts?

I could easily see this broken down into multiple blog posts like:

1. Intro and proof of concept in Jupyter notebooks.
2. Reproducibility with DVC.
3. Leveraging cloud resources with CI/CD and CML.
4. Experiment management with DVCLive and Studio.

This would also enable you to go deeper in each post. For example, you link to the code a lot but don't really show any of it. You also don't mention revisit the specifics of the use case much (like how the production app might be used or how what to look for in evaluation). Without fitting so much into one blog post, you might be able to incorporate more of these details, which could make it more generally useful and interesting.

**Comment by alex000kim** [comment]
Yeah, I totally get your point. 
The choice was between something super high-level that covers everything OR, as you suggest, a series of 3-4 posts. 
Initially, I opted for the former, but I am open to the latter if others agree.
My only concern here would be that people might not get the overall picture by not reading the entire series of posts.

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+then the answer is yes if you use [CML library](https://cml.dev/) in combination
+with CI/CD tools like [GitHub Actions](https://docs.github.com/en/actions) or
+[GitLab CI/CD](https://docs.gitlab.com/ee/ci/).
+
+We've configured three
+[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/master/.github/workflows)
+for GitHub Actions, each of which corresponds to a particular stage depending on
+the project's lifecycle we are in:
+
+1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/train-model.yaml)
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:309]:
```suggestion
1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/1-experiment.yaml)
```

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+   Now we can test our API and assess the end-to-end performance of the overall
+   solution.
+
+```mermaid
+flowchart TB
+A(Merge to dev branch) --> B(Retraining) --> C(Deployment to dev environment)
+
+```
+
+1. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/prod-deploy-api-to-heroku.yaml)
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:352]:
```suggestion
3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/3-deploy.yaml)
```

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+   At this point, we can assess the results in DVC Studio and GitHub and decide
+   what things we want to change next.
+
+```mermaid
+flowchart TB
+A(Commit to exp branch) -->|Push changes| B("Exp CML workflow\n(training & reporting)")
+B --> |Reports,\\nmetrics,\\nplots| C("Check results.\nAre they good?")
+C --> |No\\nchange experiment parameters | A
+C -->|Yes| D(PR to dev branch)
+```
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:331]: This is pending from #3300 , right ?

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+1. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/dev-train-upload-deploy.yaml)
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the dev branch. This would trigger a different CI/CD job
+   that will:
+
+   1. retrain the model with the new parameters
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:339]: If I understand the workflow correctly, dvc won't actually run the pipeline, right?

If this is True, it feels like this workflow should not launch a cloud instance nor try to run the pipeline at all, unless we are considering the possibility of someone pushing changes directly to develop?

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+   this triggers the 3rd CI/CD workflow that deploys the code from the main
+   branch to the production API.
+
+```mermaid
+flowchart TB
+B(Merge dev into main) --> C(Prod CML workflow) --> D(Deployment to prod)
+
+```
+
+### Experiment Management with DVCLive and Studio
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:365]: Again, It feels like the above section could be expanded into a separate blog post. 

I would like to see the actual code (or a simplified version) corresponding to each of the bullet points that describe what each workflow does.

**Review by daavoo — COMMENTED** [review]
🔥 Like the content. Main concern is what @dberenbaum mention about broad scope. I'm not sure about the 4 parts, I was leading more towards merging 1-2 and 3-4

File: content/blog/2022-03-24-e2e-computer-vision.md [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+[Studio](https://studio.iterative.ai/) is a web-based application with seamless
```
Comment (daavoo) [inline content/blog/2022-03-24-e2e-computer-vision.md:367]: Given that we had set up a CI/CD workflow, I would expand this section including info about the ability to perform data/params updates and Run Experiments from Studio UI https://dvc.org/doc/studio/user-guide/views/run-experiments

**Comment by daavoo** [comment]
>  My only concern here would be that people might not get the overall picture by not reading the entire series of posts.

I think this should not be an issue if at the beginning and end of each individual post you reiterate on the overall picture and link to the other parts

**Comment by alex000kim** [comment]
Thanks, @daavoo for the super detailed comments.
Unless there are other objections, I can close this PR and split this post into two:
1. POC, data versioning, ML pipelines 
2. Experiment management (DVCLive + Studio), CI/CD (CML + cloud resources) 
Most of the content will remain the same, but I can elaborate on a couple of points for clarity.

**Pull request closed without merging**